File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/utils/views/recycler_view/RecyclerViewUtils.java
package com.github.anrimian.musicplayer.ui.utils.views.recycler_view;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.view.View;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.LinearSmoothScroller;
import androidx.recyclerview.widget.RecyclerView;

import com.github.anrimian.musicplayer.R;
import com.github.anrimian.musicplayer.domain.utils.java.Callback;
import com.github.anrimian.musicplayer.domain.utils.java.CompositeCallback;
import com.github.anrimian.musicplayer.ui.utils.views.recycler_view.touch_helper.swipe_to_delete.SwipeToDeleteItemDecorator;
import com.github.anrimian.musicplayer.ui.utils.views.recycler_view.touch_helper.swipe_to_delete.SwipeToDeleteTouchHelperCallback;

import me.zhanghai.android.fastscroll.FastScroller;
import me.zhanghai.android.fastscroll.FastScrollerBuilder;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static com.github.anrimian.musicplayer.ui.utils.ViewUtils.animateVisibility;

public class RecyclerViewUtils {

    /**
     * app-specific func
     */
    public static boolean isPositionVisible(LinearLayoutManager lm, int position) {
        return position > lm.findFirstVisibleItemPosition() &&
                position < lm.findLastCompletelyVisibleItemPosition();
    }

    public static void scrollToPosition(RecyclerView rv,
                                        LinearLayoutManager lm,
                                        int position,
                                        boolean smooth) {
        Context context = rv.getContext();
        rv.post(() -> {
            if (smooth) {
                RecyclerViewUtils.smoothScrollToTop(position,
                        lm,
                        context,
                        170);
            } else {
                lm.scrollToPositionWithOffset(position, 0);
            }
        });
    }

    public static <T extends RecyclerView.ViewHolder> void viewHolders(RecyclerView recyclerView,
                                                                       Callback<T> callback) {
        for (int childCount = recyclerView.getChildCount(), i = 0; i < childCount; ++i) {
            View view = recyclerView.getChildAt(i);
            RecyclerView.ViewHolder viewHolder = recyclerView.getChildViewHolder(view);
            //noinspection unchecked
            callback.call((T) viewHolder);
        }
    }

    public static void smoothScrollToTop(int position,
                                         RecyclerView.LayoutManager layoutManager,
                                         Context context,
                                         int duration) {
        RecyclerView.SmoothScroller smoothScroller = new LinearSmoothScroller(context) {

            @Override
            protected int getVerticalSnapPreference() {
                return LinearSmoothScroller.SNAP_TO_START;
            }

            @Override
            protected int calculateTimeForScrolling(int dx) {
                return duration;
            }
        };
        smoothScroller.setTargetPosition(position);
        layoutManager.startSmoothScroll(smoothScroller);
    }

    public static void attachSwipeToDelete(RecyclerView recyclerView,
                                           @ColorInt int backgroundColor,
                                           Callback<Integer> swipeCallback) {
        CompositeCallback<Integer> compositeCallback = new CompositeCallback<>();
        compositeCallback.add(swipeCallback);
        compositeCallback.add(i -> {
            RecyclerView.ItemDecoration decoration = new SwipeToDeleteItemDecorator(backgroundColor);
            recyclerView.addItemDecoration(decoration);
            RecyclerView.ItemAnimator itemAnimator = recyclerView.getItemAnimator();
            if (itemAnimator != null) {
                recyclerView.postDelayed(() ->
                                itemAnimator.isRunning(() -> recyclerView.removeItemDecoration(decoration)),
                        itemAnimator.getRemoveDuration());
            }
        });

        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(new SwipeToDeleteTouchHelperCallback(
                backgroundColor, compositeCallback));
        itemTouchHelper.attachToRecyclerView(recyclerView);
    }

    public static void attachDynamicShadow(RecyclerView recyclerView, View shadow) {
        shadow.setVisibility(
                recyclerView.computeVerticalScrollOffset() > 0? VISIBLE: View.GONE
        );
        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                animateVisibility(
                        shadow,
                        recyclerView.computeVerticalScrollOffset() > 0? VISIBLE: GONE
                );
            }
        });
    }


    public static void attachFastScroller(RecyclerView recyclerView) {
        attachFastScroller(recyclerView, false);
    }

    //attach if high amount of elements appeared? - seems working from the box
    //reduce size
    //add elevation? or color in light theme
    //list performance on fast scrolling
    public static void attachFastScroller(RecyclerView recyclerView, boolean useFabPadding) {
        Context context = recyclerView.getContext();
        Drawable trackDrawable = ContextCompat.getDrawable(context, R.drawable.drawable_empty);
        assert trackDrawable != null;
        Drawable thumbDrawable = ContextCompat.getDrawable(context, R.drawable.a_thumb_drawable);
        assert thumbDrawable != null;

        Resources resources = context.getResources();
        int thumbVerticalPadding = resources.getDimensionPixelSize(R.dimen.scrollbar_thumb_vertical_padding);
        int listBottomPadding = useFabPadding? resources.getDimensionPixelSize(R.dimen.list_bottom_padding_with_fab) : 0;

        Rect padding = new Rect(0,
                thumbVerticalPadding + recyclerView.getPaddingTop(),
                0,
                thumbVerticalPadding + listBottomPadding);

        new FastScrollerBuilder(recyclerView)
                .setTrackDrawable(trackDrawable)
                .setThumbDrawable(thumbDrawable)
                .setPadding(padding)
                .build();
    }
}

File: shared/app/src/main/java/com/github/anrimian/musicplayer/ui/utils/views/recycler_view/RecyclerViewUtils.kt
package com.github.anrimian.musicplayer.ui.utils.views.recycler_view

import android.content.Context
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.LinearSmoothScroller
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.SmoothScroller

fun scrollToPosition(
    rv: RecyclerView,
    lm: LinearLayoutManager,
    position: Int,
    smooth: Boolean,
) {
    val context = rv.context
    rv.post{
        if (smooth) {
            smoothScrollToTop(position, lm, context, 170)
        } else {
            lm.scrollToPositionWithOffset(position, 0)
        }
    }
}

fun smoothScrollToTop(
    position: Int,
    layoutManager: RecyclerView.LayoutManager,
    context: Context,
    duration: Int,
) {
    val smoothScroller: SmoothScroller = object : LinearSmoothScroller(context) {
        override fun getVerticalSnapPreference(): Int {
            return SNAP_TO_START
        }

        override fun calculateTimeForScrolling(dx: Int): Int {
            return duration
        }
    }
    smoothScroller.targetPosition = position
    layoutManager.startSmoothScroll(smoothScroller)
}

fun windowedScrollToPosition(
    context: Context,
    layoutManager: LinearLayoutManager,
    adapter: RecyclerView.Adapter<*>,
    position: Int,
    firstVisiblePosition: Int,
    lastVisiblePosition: Int,
    scrollWindowTopOffset: Int,
    scrollWindowBottomOffset: Int,
) {
    val scrollWindowBottomPositionOffset = (lastVisiblePosition - firstVisiblePosition) / 2

    val firstSmoothScrollPosition = (firstVisiblePosition - 2).coerceAtLeast(0)
    if (position < firstSmoothScrollPosition || position > lastVisiblePosition) {
        // invisible at all, fast scroll to position

        // recycler view doesn't scroll to exact position,
        // it scrolls to make target position visible
        // so add shift to target position and scroll direction is important here
        val targetPosition = if (position < firstSmoothScrollPosition) {
            // fast scroll backward
            (position - scrollWindowBottomPositionOffset - scrollWindowBottomOffset)
                .coerceAtLeast(0)
        } else {
            // fast scroll forward
            (position + scrollWindowBottomPositionOffset - scrollWindowBottomOffset)
                .coerceAtMost(adapter.itemCount - 1)
        }
        layoutManager.scrollToPosition(targetPosition)
        return
    }

    val scrollWindowTopPosition = firstVisiblePosition + scrollWindowTopOffset
    val scrollWindowBottomPosition = firstVisiblePosition + scrollWindowBottomPositionOffset
    val targetPosition = if (position < scrollWindowTopPosition) {
        // slightly atop if scroll window case
        // smooth scroll to top window border
        (position - scrollWindowTopOffset).coerceAtLeast(0)
    } else if (position > scrollWindowBottomPosition + scrollWindowBottomOffset) {
        // visible, smooth scroll to bottom window border
        position - scrollWindowBottomPositionOffset - scrollWindowBottomOffset
    } else {
        // in scroll window borders, do nothing
        return
    }


    val scroller = object: LinearSmoothScroller(context) {
        override fun getVerticalSnapPreference(): Int {
            return SNAP_TO_START
        }

        override fun calculateTimeForScrolling(dx: Int): Int {
            return 170
        }
    }
    scroller.targetPosition = targetPosition
    layoutManager.startSmoothScroll(scroller)
}